packages/@expo/server/src/index.ts

```diff
@@ -126,6 +147,54 @@ export function createRequestHandler(
 
     debug('Request', sanitizedPathname);
 
+    // Execute middleware first, before any route matching
+    if (routesManifest.middleware) {
+      try {
+        const middlewareModule = await getMiddleware(routesManifest.middleware);
+        if (middlewareModule?.default) {
+          const middlewareResponse = await middlewareModule.default(request);
+          if (middlewareResponse instanceof Response) {
+            debug('Middleware returned response, short-circuiting');
+            return middlewareResponse;
+          }
+          // If middleware returns undefined/void, continue to route matching
+        }
+      } catch (error) {
+        // Only show detailed errors for development; in production, show a generic error to prevent leaking sensitive information
+        debug('Middleware execution error:', error);
+
+        const isApiRequest = routesManifest.apiRoutes.some((route) =>
+          route.namedRegex.test(sanitizedPathname)
+        );
+
+        if (isApiRequest) {
+          return new Response(
+            JSON.stringify({
+              error:
+                process.env.NODE_ENV === 'development'
+                  ? error
+                  : { message: 'Internal Server Error' },
+            }),
+            {
+              status: 500,
+              headers: { 'Content-Type': 'application/json' },
+            }
+          );
```

Comment:
I like that in dev the message is returned, but thinking out loud, users could base some logic on the error message, which would be unwanted.

For API Routes the message is logged and "production" error is returned.

https://github.com/expo/expo/blob/60b028cb6220b412a5af6cc3b474501ea32179d8/packages/%40expo/server/src/index.ts#L289-L293